### PR TITLE
Allow supporting merge PRs with changelogger.

### PR DESCRIPTION
**Pull request by vedanshujain**
#### Changes proposed in this Pull Request:

In WooCommerce, we are hoping to use merge strategy along with squash for merging PRs. However, Jetpack changelogger currently does not support the merge strategy when param `add-pr-num` is provided. **Edit:** Now default merge message of format "Merge pull request #23123 from branch" is also supported.

This is because, in the git-log command, we list the latest commit for the changelog file, whereas what we want is the merge commit (which contains the PR number towards the in (#{pr_no}) format. This PR add supports to list merge commit, or a squash commit, depending upon what is available for the particular change, and therefore supports both strategies.

#### Other information:

- [ ] Have you written new tests for your changes, if applicable?
- [x] Have you checked the E2E test CI results, and verified that your changes do not break them?

#### Does this pull request change what data or activity we track or use?

* No

#### Testing instructions:

1. Create a new branch, say `temp`.
2. Create another branch with `temp` as base. Say `temp2`.
3. Add couple of commits to `temp2` branch.
4. Add a changelog.
5. Switch to `temp` branch, and merge the `temp2` branch with command like so: `git merge temp2 --no-ff`. For merge commit message, use something like `Allow supporting merge PRs with changelogger. (#27881)`. This is the default message format when the following setting is enabled:
Screenshot 2022-12-12 at 3 19 47 PM
6. Run the changelogger command with params `--add-pr-num`. Check that the generated changelog has the PR number included.
7. Repeat the same, but this time in the merge commit message, use `Merge pull request #27881 from changelogger/merge` and verify that the change logger can pick the PR numbers.

### CODE REVIEW

**Comment by vedanshujain** [comment]
Marking for team review, since the failing test relates to updating change logger version, which is probably not something that I should update (although let me know if I should).

**Review by jeherve — DISMISSED** [review]
> which is probably not something that I should update (although let me know if I should).

You can definitely make that change on your own. If you run the command that's recommended in the failing check, it should fix things for you.

Thank you!

**Comment by vedanshujain** [comment]
hey @jeherve, seems like currently, the CI failing since I have had to update changelogger version for all packages, and now it is asking me to add changelog files for all of them. before going ahead just wanted to confirm if this is expected, shouldn't the version update be happening after the pr is merged and a new version is available? is there a way to add changelog for all packages at once or to ignore the check (since the update is in dev dependency and nothing is really being changed).

**Comment by jeherve** [comment]
> just wanted to confirm if this is expected

It is indeed. Since you're the first to make changes to the package since its last release, and since all plugins in this repo rely on the changelogger package, each one of those plugins need to be updated to use the right version of the changelogger package.

Luckily, you can rely on the `jetpack` CLI tool to bulk create changelog entries where needed. Type `jetpack changelog add` and follow the prompts. That should help.

**Comment by vedanshujain** [comment]
Thanks @jeherve , should be ready for review now.

**Comment by anomiex** [comment]
> whereas what we want is the merge commit (which contains the PR number towards the in (#{pr_no}) format.

That's not what GitHub does for merge commits. The commit message instead would look something like "Merge pull request #27881 from changelogger/merge".

**Comment by vedanshujain** [comment]
> The commit message instead would look something like "Merge pull request https://github.com/Automattic/jetpack/pull/27881 from changelogger/merge".


Right, this is true for the default commit message, but you can configure and set it to be similar to squash by going to repository settings like so (which I think is what we want to go for in the WC repo):

Screenshot 2022-12-14 at 3 50 46 PM

See this [merge commit](https://github.com/vedanshujain/woocommerce/commit/a3d12954e9f8bb62aa447dfcf68754dd5b7c0c74) for example. 

I can add support for the default setting given that the default setting is configurable (although it would need to wait at least a week).

**Comment by anomiex** [comment]
Interesting, I didn't know that was configurable.

I do think it would be best to support the default setting too.

**Comment by vedanshujain** [comment]
@anomiex is ready for review again (e2e failure looks unrelated and is happening on at least one other recent PR (https://github.com/Automattic/jetpack/actions/runs/3730708345/jobs/6328086286)).

File: projects/packages/changelogger/src/Utils.php
Comment (anomiex) [inline projects/packages/changelogger/src/Utils.php]: We don't need so many capturing groups in the regex. OTOH, we should probably keep the anchoring.

Also, `empty()` already incorporates `isset()` logic, there's no need to use both here.
```suggestion
					preg_match( '/^Merge pull request #(\d+)|\(#(\d+)\)$/', $cmd_output[1], $matches );
					if ( ! empty( $matches[1] ) ) {
						$repo_data['pr-num'] = $matches[1];
					} elseif ( ! empty( $matches[2] ) ) {
						$repo_data['pr-num'] = $matches[2];
```

**Review by anomiex — COMMENTED** [review]
One more thing, please add test coverage for this change.

You could do this in [tests/php/tests/src/UtilsTest.php](https://github.com/Automattic/jetpack/blob/cb5cfb0993cfd23806356bc143bc584391ac9de8/projects/packages/changelogger/tests/php/tests/src/UtilsTest.php#L354). You'd probably want to add `-b main` or the like to [the `git init` command](https://github.com/Automattic/jetpack/blob/cb5cfb0993cfd23806356bc143bc584391ac9de8/projects/packages/changelogger/tests/php/tests/src/UtilsTest.php#L392), then add another file with git commands like this to simulate a GitHub merge commit
```
git checkout -b branch/whatever
git add in-git-merge.txt
git commit -m 'Commit on branch'
git checkout main
git merge --no-ff -m 'Merge pull request #567 from branch/whatever' branch/whatever
```

**Comment by vedanshujain** [comment]
@anomiex addressed both the changes, let me know what you think!

File: projects/packages/changelogger/src/Utils.php
Comment (vedanshujain) [reply projects/packages/changelogger/src/Utils.php]: Added in https://github.com/Automattic/jetpack/pull/27881/commits/4940548c2ca50db539244eb52fa3ddd784a814ce

File: projects/packages/changelogger/tests/php/tests/src/UtilsTest.php
Comment (anomiex) [inline projects/packages/changelogger/tests/php/tests/src/UtilsTest.php]: Personally I would now get rid of the `testGetRepoData` test above and just have this new test (renamed to `testGetRepoData`), as it tests all the exact same things plus the new code.

**Review by anomiex — DISMISSED** [review]
Thanks! Looks good to me; I left one suggestion inline, but feel free to ignore it if you don't like it.

File: projects/packages/changelogger/tests/php/tests/src/UtilsTest.php
Comment (vedanshujain) [reply projects/packages/changelogger/tests/php/tests/src/UtilsTest.php]: makes sense, addressed in https://github.com/Automattic/jetpack/pull/27881/commits/1c43fa3c081c55f017dccbd538655a12d917dc3e

**Comment by vedanshujain** [comment]
@anomiex , seems like i need another review, hopefully the last one for this pr :)

**Review by anomiex — APPROVED** [review]
Let's do it!

**Pull request merged**
